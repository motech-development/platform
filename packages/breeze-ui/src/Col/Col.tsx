import React, { FC, memo, ReactNode } from 'react';
import styled from 'styled-components';

interface IColContainer {
  align: 'centre' | 'left' | 'right';
  lg: number;
  lgOffset: number;
  md: number;
  mdOffset: number;
  sm: number;
  smOffset: number;
  verticalAlign: boolean;
  xs: number;
  xsOffset: number;
}

const ColContainer = styled.div<IColContainer>`
  ${({
    align,
    lg,
    lgOffset,
    md,
    mdOffset,
    sm,
    smOffset,
    verticalAlign,
    xs,
    xsOffset,
  }) => `
    grid-column: ${xsOffset > 0 ? `${xsOffset} / span ${xs}` : `span ${xs}`};
    position: relative;
    text-align ${align === 'centre' ? 'center' : align};

    ${
      verticalAlign
        ? `
      align-items: center;
      display: flex;
    `
        : ``
    }

    ${
      sm
        ? `
      @media (min-width: 576px) {
        grid-column: ${
          smOffset > 0 ? `${smOffset} / span ${sm}` : `span ${sm}`
        };
      }
    `
        : ''
    }

    ${
      md
        ? `
      @media (min-width: 768px) {
        grid-column: ${
          mdOffset > 0 ? `${mdOffset} / span ${md}` : `span ${md}`
        };
      }
    `
        : ''
    }

    ${
      lg
        ? `
      @media (min-width: 992px) {
        grid-column: ${
          lgOffset > 0 ? `${lgOffset} / span ${lg}` : `span ${lg}`
        };
      }
    `
        : ''
    }
  `}
`;

const VericalAlignWrapper = styled.div`
  flex: 1;
`;

export interface IColProps {
  children: ReactNode;
  align?: 'centre' | 'left' | 'right';
  lg?: number;
  lgOffset?: number;
  md?: number;
  mdOffset?: number;
  sm?: number;
  smOffset?: number;
  verticalAlign?: 'middle';
  xs?: number;
  xsOffset?: number;
}

const Col: FC<IColProps> = ({
  children,
  align = 'left',
  lg = 0,
  lgOffset = 0,
  md = 0,
  mdOffset = 0,
  sm = 0,
  smOffset = 0,
  verticalAlign = null,
  xs = 12,
  xsOffset = 0,
}) => (
  <ColContainer
    align={align}
    lg={lg}
    lgOffset={lgOffset}
    md={md}
    mdOffset={mdOffset}
    sm={sm}
    smOffset={smOffset}
    verticalAlign={!!verticalAlign}
    xs={xs}
    xsOffset={xsOffset}
  >
    {verticalAlign ? (
      <VericalAlignWrapper>{children}</VericalAlignWrapper>
    ) : (
      children
    )}
  </ColContainer>
);

export default memo(Col);
