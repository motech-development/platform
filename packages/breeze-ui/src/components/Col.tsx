import { FC, ReactNode } from 'react';
import TAlignment from '../utils/alignment';
import { classNames, textAlignmentClass } from '../utils/className';
import { TColumn } from '../utils/grid';

export interface IColProps {
  children: ReactNode;
  align?: TAlignment;
  lg?: TColumn;
  lgOffset?: TColumn;
  md?: TColumn;
  mdOffset?: TColumn;
  sm?: TColumn;
  smOffset?: TColumn;
  verticalAlign?: boolean;
  xs?: TColumn;
  xsOffset?: TColumn;
}

const colClasses = (size: string, cols: number, offset: number) => {
  if (cols > 0) {
    if (offset > 0) {
      return `${size}:col-start-${offset} ${size}:col-span-${cols}`;
    }

    return `${size}:col-span-${cols}`;
  }

  return '';
};

const Col: FC<IColProps> = ({
  children,
  align = 'left',
  lg = 0,
  lgOffset = 0,
  md = 0,
  mdOffset = 0,
  sm = 0,
  smOffset = 0,
  verticalAlign = false,
  xs = 12,
  xsOffset = 0,
}) => (
  <>
    {/* @tailwind: col-start-1 col-start-2 col-start-3 col-start-4 col-start-5 col-start-6 col-start-7 col-start-8 col-start-9 col-start-10 col-start-11 col-start-12 */}
    {/* @tailwind: col-span-1 col-span-2 col-span-3 col-span-4 col-span-5 col-span-6 col-span-7 col-span-8 col-span-9 col-span-10 col-span-11 col-span-12 */}
    {/* @tailwind: sm:col-start-1 sm:col-start-2 sm:col-start-3 sm:col-start-4 sm:col-start-5 sm:col-start-6 sm:col-start-7 sm:col-start-8 sm:col-start-9 sm:col-start-10 sm:col-start-11 sm:col-start-12 */}
    {/* @tailwind: sm:col-span-1 sm:col-span-2 sm:col-span-3 sm:col-span-4 sm:col-span-5 sm:col-span-6 sm:col-span-7 sm:col-span-8 sm:col-span-9 sm:col-span-10 sm:col-span-11 sm:col-span-12 */}
    {/* @tailwind: md:col-start-1 md:col-start-2 md:col-start-3 md:col-start-4 md:col-start-5 md:col-start-6 md:col-start-7 md:col-start-8 md:col-start-9 md:col-start-10 md:col-start-11 md:col-start-12 */}
    {/* @tailwind: md:col-span-1 md:col-span-2 md:col-span-3 md:col-span-4 md:col-span-5 md:col-span-6 md:col-span-7 md:col-span-8 md:col-span-9 md:col-span-10 md:col-span-11 md:col-span-12 */}
    {/* @tailwind: lg:col-start-1 lg:col-start-2 lg:col-start-3 lg:col-start-4 lg:col-start-5 lg:col-start-6 lg:col-start-7 lg:col-start-8 lg:col-start-9 lg:col-start-10 lg:col-start-11 lg:col-start-12 */}
    {/* @tailwind: lg:col-span-1 lg:col-span-2 lg:col-span-3 lg:col-span-4 lg:col-span-5 lg:col-span-6 lg:col-span-7 lg:col-span-8 lg:col-span-9 lg:col-span-10 lg:col-span-11 lg:col-span-12 */}
    <div
      className={classNames(
        'relative',
        textAlignmentClass(align),
        verticalAlign ? 'flex items-center' : '',
        xsOffset > 0
          ? `col-start-${xsOffset} col-span-${xs}`
          : `col-span-${xs}`,
        colClasses('sm', sm, smOffset),
        colClasses('md', md, mdOffset),
        colClasses('lg', lg, lgOffset),
      )}
    >
      <div className={classNames(verticalAlign ? 'flex-1' : '')}>
        {children}
      </div>
    </div>
  </>
);

export default Col;
