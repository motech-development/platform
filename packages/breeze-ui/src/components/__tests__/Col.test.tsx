import { render } from '@testing-library/react';
import Col from '../Col';

describe('Col', () => {
  describe('xs', () => {
    it('should have the correct default styles', () => {
      const { container } = render(<Col>Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct styles when size is set', () => {
      const { container } = render(<Col xs={6}>Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct styles when size and offset is set', () => {
      const { container } = render(
        <Col xs={6} xsOffset={4}>
          Hello
        </Col>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('sm', () => {
    it('should have the correct styles when size is set', () => {
      const { container } = render(<Col sm={6}>Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct styles when size and offset is set', () => {
      const { container } = render(
        <Col sm={6} smOffset={4}>
          Hello
        </Col>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('md', () => {
    it('should have the correct styles when size is set', () => {
      const { container } = render(<Col md={6}>Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct styles when size and offset is set', () => {
      const { container } = render(
        <Col md={6} mdOffset={4}>
          Hello
        </Col>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('lg', () => {
    it('should have the correct styles when size is set', () => {
      const { container } = render(<Col lg={6}>Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct styles when size and offset is set', () => {
      const { container } = render(
        <Col lg={6} lgOffset={4}>
          Hello
        </Col>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('alignment', () => {
    it('should have the correct styles when alignment set to left', () => {
      const { container } = render(<Col align="left">Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct styles when alignment set to right', () => {
      const { container } = render(<Col align="right">Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct styles when alignment set to centre', () => {
      const { container } = render(<Col align="center">Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should have the correct styles when vertical alignment set to middle', () => {
      const { container } = render(<Col verticalAlign>Hello</Col>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
