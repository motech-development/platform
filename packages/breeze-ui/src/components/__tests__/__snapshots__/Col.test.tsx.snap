// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col alignment should have the correct styles when alignment set to centre 1`] = `
<div
  class="relative text-center col-span-12"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col alignment should have the correct styles when alignment set to left 1`] = `
<div
  class="relative text-left col-span-12"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col alignment should have the correct styles when alignment set to right 1`] = `
<div
  class="relative text-right col-span-12"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col alignment should have the correct styles when vertical alignment set to middle 1`] = `
<div
  class="relative text-left flex items-center col-span-12"
>
  <div
    class="flex-1"
  >
    Hello
  </div>
</div>
`;

exports[`Col lg should have the correct styles when size and offset is set 1`] = `
<div
  class="relative text-left col-span-12 lg:col-start-4 lg:col-span-6"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col lg should have the correct styles when size is set 1`] = `
<div
  class="relative text-left col-span-12 lg:col-span-6"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col md should have the correct styles when size and offset is set 1`] = `
<div
  class="relative text-left col-span-12 md:col-start-4 md:col-span-6"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col md should have the correct styles when size is set 1`] = `
<div
  class="relative text-left col-span-12 md:col-span-6"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col sm should have the correct styles when size and offset is set 1`] = `
<div
  class="relative text-left col-span-12 sm:col-start-4 sm:col-span-6"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col sm should have the correct styles when size is set 1`] = `
<div
  class="relative text-left col-span-12 sm:col-span-6"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col xs should have the correct default styles 1`] = `
<div
  class="relative text-left col-span-12"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col xs should have the correct styles when size and offset is set 1`] = `
<div
  class="relative text-left col-start-4 col-span-6"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;

exports[`Col xs should have the correct styles when size is set 1`] = `
<div
  class="relative text-left col-span-6"
>
  <div
    class=""
  >
    Hello
  </div>
</div>
`;
